### PR TITLE
[PM-33848] feat: Add dismissable Premium Upgrade banner to vault list

### PR DESCRIPTION
**Pull request by andrebispo5**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-33848

## 📔 Objective

Add a dismissable "Upgrade to Premium" banner to the vault list view, gated by the existing `premiumUpgradePath` feature flag.

The banner appears for free users when:
- Account is 7+ days old
- Vault has 5+ items
- Banner has not been previously dismissed

Changes include:
- Add `shouldShowPremiumUpgradeBanner()` method to StateService
- Add `hasMinimumCipherCount(_:)` method to VaultRepository
- Add premium upgrade ActionCard to VaultListView
- Add localization strings for banner title and message
- Add unit tests for all new functionality

Note: The "Upgrade to Premium" button action is a placeholder (TODO: PM-33849) for future navigation implementation.

## 📸 Screenshots

Simulator Screenshot - iPhone 17 Pro - 2026-03-31 at 16 56 42

### CODE REVIEW

File: BitwardenResources/Localizations/en.lproj/Localizable.strings [hunk 1]
```diff
@@ -658,6 +658,8 @@
 "Totp" = "TOTP";
 "VerificationCodes" = "Verification codes";
 "PremiumSubscriptionRequired" = "Premium subscription required";
+"PremiumPlanGivesMoreTools" = "A Premium plan gives you more tools to stay secure and in control.";
```
Comment (KatherineInCode) [inline BitwardenResources/Localizations/en.lproj/Localizable.strings]: ⛏️ We try to match up as exactly as we can with the keys, along with appending `DescriptionLong` if we don't get everything. I would probably put this one as `APremiumPlanGivesYouMoreToolsDescriptionLong`.

File: BitwardenResources/Localizations/en.lproj/Localizable.strings [hunk 2]
```diff
@@ -1330,4 +1332,12 @@
 "LearnMoreAboutAutofill" = "Learn more about autofill";
 "TurnOnNow" = "Turn on now";
 "AcceptTransfer" = "Accept transfer";
+"PerMonth" = " / month";
```
Comment (KatherineInCode) [inline BitwardenResources/Localizations/en.lproj/Localizable.strings]: 🤔 I'm not sure how much I love putting the division punctuation in like this

File: BitwardenResources/Localizations/en.lproj/Localizable.strings [hunk 3]
```diff
@@ -1330,4 +1332,12 @@
...
+"UnlockMoreAdvancedFeaturesWithPremiumPlan" = "Unlock more advanced features with a Premium plan.";
+"BuiltInAuthenticator" = "Built-in authenticator";
+"EmergencyAccess" = "Emergency access";
+"SecureFileStorage" = "Secure file storage";
+"BreachMonitoring" = "Breach monitoring";
+"UpgradeNow" = "Upgrade now";
+"YouWillGoToStripeSecureCheckout" = "You'll go to Stripe's secure checkout to complete your purchase.";
```
Comment (KatherineInCode) [inline BitwardenResources/Localizations/en.lproj/Localizable.strings]: ⛏️ 
```suggestion
"YoullGoToStripesSecureCheckoutDescriptionLong" = "You’ll go to Stripe's secure checkout to complete your purchase.";
```

We admittedly don't have a standard for how to handle contractions (though I feel like we've discussed it before), but the majority of keys just elide the apostrophe, so we'll want to do that and `DescriptionLong`. As well, we should use a proper curly apostrophe `’` in the English

File: BitwardenShared/Core/Platform/Services/ServiceContainer.swift [hunk 4]
```diff
@@ -27,6 +27,9 @@ public class ServiceContainer: Services { // swiftlint:disable:this type_body_le
     /// The service used by the application to make API requests.
     let apiService: APIService
 
+    /// An optional override for the billing API service, used for testing.
+    var billingAPIServiceOverride: BillingAPIService?
```
Comment (KatherineInCode) [inline BitwardenShared/Core/Platform/Services/ServiceContainer.swift]: ⛏️ This should be alphabetized

File: BitwardenShared/Core/Platform/Services/StateService.swift [hunk 5]
```diff
@@ -1351,6 +1351,23 @@ extension StateService {
...
+    func shouldShowPremiumUpgradeBanner() async -> Bool {
+        guard await !doesActiveAccountHavePremium() else { return false }
+
+        let dismissed = await ((try? getPremiumUpgradeBannerDismissed()) ?? false)
+        guard !dismissed else { return false }
+
+        // Check account age >= 7 days
+        guard let account = try? await getActiveAccount(),
+              let creationDate = account.profile.creationDate else { return false }
+        let sevenDays: TimeInterval = 7 * 24 * 60 * 60
```
Comment (KatherineInCode) [inline BitwardenShared/Core/Platform/Services/StateService.swift]: 🎨 I would recommend putting this into one of our [Constants.swift files](https://github.com/bitwarden/ios/blob/main/BitwardenShared/Core/Platform/Utilities/Constants.swift), since that's where we keep these sorts of things already

File: BitwardenShared/Core/Platform/Services/StateService.swift [hunk 6]
```diff
@@ -1351,6 +1351,23 @@ extension StateService {
...
+        guard Date().timeIntervalSince(creationDate) >= sevenDays else { return false }
```
Comment (KatherineInCode) [inline BitwardenShared/Core/Platform/Services/StateService.swift]: 🤔 I always get nervous with a raw `Date()`; would it maybe make sense to bring in the `TimeProvider` here? It's probably six of one half dozen of the other, but it would make doing tests potentially easier

File: BitwardenShared/Core/Vault/Repositories/VaultRepository.swift [hunk 7]
```diff
@@ -714,6 +721,10 @@ extension DefaultVaultRepository: VaultRepository {
         return totp
     }
 
+    func hasMinimumCipherCount(_ count: Int) async throws -> Bool {
```
Comment (KatherineInCode) [inline BitwardenShared/Core/Vault/Repositories/VaultRepository.swift:724]: 🤔 I don't love the name of this function, but nothing better comes to mind

File: BitwardenShared/Core/Vault/Repositories/VaultRepositoryTests.swift [hunk 8]
```diff
@@ -1135,6 +1135,35 @@ class VaultRepositoryTests: BitwardenTestCase { // swiftlint:disable:this type_b
...
+    /// `hasMinimumCipherCount(_:)` returns `true` when vault has at least the specified count.
+    func test_hasMinimumCipherCount_true() async throws {
+        cipherService.cipherCountResult = .success(5)
+        let hasMinimum = try await subject.hasMinimumCipherCount(5)
+        XCTAssertTrue(hasMinimum)
+    }
+
+    /// `hasMinimumCipherCount(_:)` returns `true` when vault has more than the specified count.
+    func test_hasMinimumCipherCount_moreThanMinimum() async throws {
+        cipherService.cipherCountResult = .success(10)
```
Comment (KatherineInCode) [inline BitwardenShared/Core/Vault/Repositories/VaultRepositoryTests.swift]: 🎨 From a testing perspective, I'd rather check the boundary cases, so `6` here and `4` for the fewer case. In practice the code probably isn't going to go off-spec, but the test also could pass if it checks against `count - 1`, for instance.

File: BitwardenShared/UI/Billing/PremiumUpgrade/PremiumUpgradeCoordinatorTests.swift [hunk 9]
```diff
@@ -0,0 +1,55 @@
+import BitwardenKitMocks
+import SwiftUI
+import XCTest
+
+@testable import BitwardenShared
+@testable import BitwardenSharedMocks
```
Comment (KatherineInCode) [inline BitwardenShared/UI/Billing/PremiumUpgrade/PremiumUpgradeCoordinatorTests.swift]: 🤔 I'm pretty sure you don't need to `@testable import` the mocks, and can just `import` them directly

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessor.swift [hunk 10]
```diff
@@ -109,6 +109,9 @@ final class VaultListProcessor: StateProcessor<
             await dismissFlightRecorderToastBanner()
         case .dismissImportLoginsActionCard:
             await setImportLoginsProgress(.setUpLater)
+        case .dismissPremiumUpgradeActionCard:
+            state.shouldShowPremiumUpgradeActionCard = false
+            try? await services.stateService.setPremiumUpgradeBannerDismissed(true)
```
[the remark this answers is not in the document]
Comment (KatherineInCode) [reply BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessor.swift]: I agree that we should at the least record the error in the error reporter

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessor.swift [hunk 11]
```diff
@@ -236,6 +241,14 @@ extension VaultListProcessor {
         if await services.configService.getFeatureFlag(.archiveVaultItems) {
             state.shouldShowArchiveOnboardingActionCard = await services.stateService.shouldDoArchiveOnboarding()
         }
+
+        if await services.configService.getFeatureFlag(.premiumUpgradePath) {
+            let shouldShow = await services.stateService.shouldShowPremiumUpgradeBanner()
+            let hasEnoughItems = await (try? services.vaultRepository.hasMinimumCipherCount(5)) ?? false
```
Comment (KatherineInCode) [inline BitwardenShared/UI/Vault/Vault/VaultList/VaultListProcessor.swift]: 🎨 I think this minimum cipher count of 5 should be pulled into [Constants.swift](https://github.com/bitwarden/ios/blob/main/BitwardenShared/Core/Platform/Utilities/Constants.swift)

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListState.swift [hunk 12]
```diff
@@ -87,6 +87,9 @@ struct VaultListState: Equatable {
     /// Whether the Archive Onboarding action card should be shown.
     var shouldShowArchiveOnboardingActionCard: Bool = false
 
+    /// Whether the Premium Upgrade action card should be shown.
+    var shouldShowPremiumUpgradeActionCard: Bool = false
```
Comment (KatherineInCode) [inline BitwardenShared/UI/Vault/Vault/VaultList/VaultListState.swift]: ⛏️ Alphabetization

File: BitwardenShared/UI/Vault/Vault/VaultCoordinator.swift [hunk 13]
```diff
@@ -409,6 +412,15 @@ final class VaultCoordinator: Coordinator, HasStackNavigator { // swiftlint:disa
         stackNavigator?.present(navigationController)
     }
 
+    /// Shows the premium upgrade screen.
+    ///
+    private func showPremiumUpgrade() {
```
Comment (KatherineInCode) [inline BitwardenShared/UI/Vault/Vault/VaultCoordinator.swift]: 🤔 This looks like it presents the premium upgrade screen, but the PR description indicates that the button does nothing. It's a small thing, but if this PR will also allow for showing the screen, it should include a screenshot in the description; but if this PR will not allow for showing the screen, the code around the screen shouldn't be in the PR, in my opinion

File: BitwardenShared/UI/Vault/Vault/VaultRoute.swift [hunk 14]
```diff
@@ -77,6 +77,9 @@ public enum VaultRoute: Equatable, Hashable {
     /// A route to the vault item selection screen.
     case vaultItemSelection(TOTPKeyModel)
 
+    /// A route to the premium upgrade view.
+    case premiumUpgrade
```
Comment (KatherineInCode) [inline BitwardenShared/UI/Vault/Vault/VaultRoute.swift]: ⛏️ Alphabetization

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListView.swift [hunk 15]
```diff
@@ -268,6 +284,8 @@ private struct SearchableVaultListView: View {
     @ViewBuilder
     private func vaultContents(with sections: [VaultListSection]) -> some View {
         VStack(spacing: 20) {
+            premiumUpgradeActionCard
```
Comment (KatherineInCode) [inline BitwardenShared/UI/Vault/Vault/VaultList/VaultListView.swift:287]: 🤔 Is it possible that we might meet the criteria to show both cards? If so, do we actually want to show both cards, or should we have some sort of "stack" system so we only show one at a time?

File: BitwardenShared/UI/Vault/Vault/VaultCoordinator.swift (hunk 13 above)
Comment (andrebispo5) [reply BitwardenShared/UI/Vault/Vault/VaultCoordinator.swift]: You are totally right, I pushed these commits by mistake into this branch instead of the PremiumUpgrade UI one. I've already reverted the changes. Thank you for noticing. 👍

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListView.swift (hunk 15 above)
Comment (andrebispo5) [reply BitwardenShared/UI/Vault/Vault/VaultList/VaultListView.swift:287]: That is a good question, I'll reach out to product to find out how this should behave.

File: BitwardenResources/Localizations/en.lproj/Localizable.strings (hunk 2 above)
Comment (andrebispo5) [reply BitwardenResources/Localizations/en.lproj/Localizable.strings]: removed from this PR

File: BitwardenResources/Localizations/en.lproj/Localizable.strings (hunk 3 above)
Comment (andrebispo5) [reply BitwardenResources/Localizations/en.lproj/Localizable.strings]: removed from this PR, but will take into account in the new one.

File: BitwardenShared/Core/Platform/Services/StateService.swift (hunk 6 above)
Comment (andrebispo5) [reply BitwardenShared/Core/Platform/Services/StateService.swift]: Injected TimeProvider 👍

File: BitwardenShared/UI/Billing/PremiumUpgrade/PremiumUpgradeCoordinatorTests.swift (hunk 9 above)
Comment (andrebispo5) [reply BitwardenShared/UI/Billing/PremiumUpgrade/PremiumUpgradeCoordinatorTests.swift]: removed from this PR

File: BitwardenShared/UI/Vault/Vault/VaultList/VaultListView.swift (hunk 15 above)
Comment (andrebispo5) [reply BitwardenShared/UI/Vault/Vault/VaultList/VaultListView.swift:287]: Ok, so far they won't show on each other because they are mutually exclusive by being premium user or not.

File: BitwardenShared/Core/Platform/Utilities/Constants.swift [hunk 16]
```diff
@@ -84,6 +87,9 @@ extension Constants {
     /// The default number of KDF iterations to perform.
     static let pbkdf2Iterations = 600_000
 
+    /// The minimum account age in days required to show the premium upgrade banner.
+    static let premiumUpgradeBannerAccountAgeDays: TimeInterval = 7 * 24 * 60 * 60
```
Comment (KatherineInCode) [inline BitwardenShared/Core/Platform/Utilities/Constants.swift]: 🤔 I think there's a mismatch on the name/documentation. While it's seven days, the actual variable is in seconds (604,800)

**Review by KatherineInCode — COMMENTED** [review]
Looks good, just one small thing

File: BitwardenShared/Core/Platform/Utilities/Constants.swift (hunk 16 above)
Comment (andrebispo5) [reply BitwardenShared/Core/Platform/Utilities/Constants.swift]: Oh I see 👍

**Review by KatherineInCode — APPROVED** [review]

**Comment by andrebispo5** [comment]
@KatherineInCode Thank you 🙂
